cpp/src/arrow/python/inference.cc

```diff
@@ -332,6 +329,13 @@ class TypeInferrer {
       ++int_count_;
     } else if (PyDateTime_Check(obj)) {
       ++timestamp_micro_count_;
+      OwnedRef tzinfo(PyObject_GetAttrString(obj, "tzinfo"));
+      if (tzinfo.obj() != nullptr && tzinfo.obj() != Py_None && timezone_.empty()) {
+        // From public docs on array construction
+        // "Localized timestamps will currently be returned as UTC "
+        //     representation). "
```

Comment:
Does this mean Arrow simply stores an erroneous value? We don't do timezone conversion in Arrow, right?

Reply:
I'm actually not sure what it was intended to mean.  This was my best guess.  Not accounting for timezones info seems like a bug?  Would you prefer I try to capture the first time zone encountered as a string? Or is the preference not to have the logic in this PR in arrow in the first place?

Reply:
If I reuse the nomenclature from the test, I get without this PR (it's 14h21 UTC currently):
```python
>>> now_utc                                                                                                     
datetime.datetime(2020, 7, 20, 14, 21, 42, 96119, tzinfo=<UTC>)
>>> arr = pa.array([now_utc])                                                                                   
>>> arr                                                                                                         
<pyarrow.lib.TimestampArray object at 0x7f44b0bfcc90>
[
  2020-07-20 14:21:42.096119
]
>>> arr.type.tz                                                                                                 
>>> arr.to_pylist()                                                                                             
[datetime.datetime(2020, 7, 20, 14, 21, 42, 96119)]
>>> arr.to_pandas()                                                                                             
0   2020-07-20 14:21:42.096119
dtype: datetime64[ns]
```
```python
>>> now_with_tz                                                                                                           
datetime.datetime(2020, 7, 20, 10, 21, 42, 96119, tzinfo=<DstTzInfo 'US/Eastern' EDT-1 day, 20:00:00 DST>)
>>> arr = pa.array([now_with_tz])                                                                                         
>>> arr.type.tz                                                                                                           
>>> arr.to_pylist()                                                                                                       
[datetime.datetime(2020, 7, 20, 10, 21, 42, 96119)]
>>> arr.to_pylist()[0].tzinfo                                                                                             
>>> arr.to_pandas()                                                                                                       
0   2020-07-20 10:21:42.096119
dtype: datetime64[ns]
```
So without the PR, there's a problem for non-UTC timestamps which lose the timezone information. It seems to get worse with this PR because non-UTC timestamps get tagged as UTC without being corrected for the timezone's offset, which is misleading. At least originally you may be alerted by the absence of a timezone on the type (in Python terms, it's a "naive" timestamp).

Reply:
> It seems to get worse with this PR because non-UTC timestamps get tagged as UTC without being corrected for the timezone's offset, which is misleading.

That is not the intent of the PR, right now everything gets converted to UTC but in my environment UTC tzinfo is maintained..  I can't replicate your results above (timezone information for me is propagaged as an example with now_with_tz:

```
>>> now_with_tz = datetime.datetime(2020, 7, 20, 10, 21, 42, 96119, tzinfo=pytz.timezone('US/Eastern'))
>>> arr = pa.array([now_with_tz]) 
>>> arr.type.tz  
'UTC'
>>> arr.to_pylist() 
[datetime.datetime(2020, 7, 20, 15, 17, 42, 96119, tzinfo=<UTC>)]
>>> arr.to_pylist()[0].tzinfo
<UTC>
>>> arr.to_pandas()
0   2020-07-20 15:17:42.096119+00:00
dtype: datetime64[ns, UTC]
```